### PR TITLE
[ISSUE #3859] enhance the cal of latency for putting message

### LINKED ISSUE #3859: enhance the cal of latency for putting message
**FEATURE REQUEST**

1. Please describe the feature you are requesting.
Now the message putting latency is only counted in the following ranges：
```java
"[<=0ms]", "[0~10ms]", "[10~50ms]", "[50~100ms]", "[100~200ms]", "[200~500ms]", "[500ms~1s]", "[1~2s]", "[2~3s]", "[3~4s]", "[4~5s]", "[5~10s]", "[10s~]",
```
- This granularity is too coarse, so we can't calculate the exact latency of putting msg with those value, such as P99, P999.

2. Provide any additional detail on your proposed use case for this feature.
- Use the bucket method to calculate the write latency of P99 and P999 with a finer granularity.
计算写入延时
- The spacing of the buckets will increase with time
- We plan to use the following spacing：
image
3. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?
nice-to-have
4. If there are some sub-tasks using -[] for each subtask and create a corresponding issue to map to the sub task:

### PR DESCRIPTION
**Pull request by cserwen**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

Close #3859

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by cserwen** [comment]
Test result:
```
2022-02-18 16:27:00 INFO StoreStatsService - findPutMessageEntireTimePX 0.99=2.87ms cost 0ms
2022-02-18 16:27:00 INFO StoreStatsService - findPutMessageEntireTimePX 0.999=3.99ms cost 0ms
2022-02-18 16:27:00 INFO StoreStatsService - [PAGECACHERT] TotalPut 499371, PutMessageDistributeTime [<=0ms]:9889 [0~10ms]:489347 [10~50ms]:121 [50~100ms]:14 [100~200ms]:0 [200~500ms]:0 [500ms~1s]:0 [1~2s]:0 [2~3s]:0 [3~4s]:0 [4~5s]:0 [5~10s]:0 [10s~]:0 
```

File: store/src/main/java/org/apache/rocketmq/store/StoreStatsService.java [hunk 1]
```diff
@@ -39,6 +43,12 @@
         "[<=0ms]", "[0~10ms]", "[10~50ms]", "[50~100ms]", "[100~200ms]", "[200~500ms]", "[500ms~1s]", "[1~2s]", "[2~3s]", "[3~4s]", "[4~5s]", "[5~10s]", "[10s~]",
     };
 
+    //The rule to define buckets
+    private static final Map<Integer, Integer> PUT_MESSAGE_ENTIRE_TIME_BUCKETS = new TreeMap<>();
+    //buckets
+    private TreeMap<Long/*bucket*/, AtomicLong/*times*/> buckets = new TreeMap<>();
```
Comment (HScarb) [inline store/src/main/java/org/apache/rocketmq/store/StoreStatsService.java:49]: Consider using LongAdder to improve the performance

Comment (cserwen) [reply store/src/main/java/org/apache/rocketmq/store/StoreStatsService.java:49]: OK. I will update it.

**Comment by cserwen** [comment]
> Will this decrease the performance of putting message?

@HScarb 
I don't think it will decrease performance. The same method as before was used to count the writing latency, but a finer granularity was used. Related to the previous, just added an add() operation when msg is put.

**Comment by ni-ze** [comment]
image
Looks like result incorrect.

**Comment by ni-ze** [comment]
Repeated px99/px999, calculation result is correct.

**Review by ni-ze — APPROVED** [review]

**Review by HScarb — APPROVED** [review]

**Review by duhenglucky — APPROVED** [review]

**Pull request merged**